File: Serialize and deserialize a binary tree.cpp
class Solution
{
    public:
    //Function to serialize a tree and return a list containing nodes of tree.
      vector<int> array;
    vector<int> serialize(Node *root) //storing elements in array using inorder traversal
    {
        //Your code here
      if(root==NULL){
          return array;
      }
          serialize(root->left);
          if(root->data>0 and root->data<=1000000009)
          array.push_back(root->data);
          serialize(root->right);
      return array;
    }
    
    //Function to deserialize a list and construct the tree.
    Node * deSerialize(vector<int> &A)
    {
       //Your code here
       Node* root = NULL;
       int l = A.size();
       
      for(int i=0;i<l;i+=3){ //creating tree using inorder traversal
           Node* lft;
           Node* rt;
           Node* ryt;
           
          if(root==NULL){
          lft = new Node(A[i]);
          root=lft;
          }
          else{
              lft = new Node(A[i]);
             Node* temp =  lft;
             lft->left = root;
             root = lft;
          }
          
          if(i+1 < l){
           rt = new Node(A[i+1]);
              rt->left=lft;
              root = rt;
              
          }if(i+2 < l){
          ryt = new Node(A[i+2]);
          root->right=ryt;
          }
          
      }
       
       return root;
    }

};

File: Serialize and deserialize a binary tree.java
class Tree
{
    
    // Function to serialize a tree and return a list containing nodes of tree.
    public void inOrder(Node root, ArrayList<Integer>list)
    {
        if(root == null)
        return;
        
        inOrder(root.left, list);
        list.add(root.data);
        inOrder(root.right, list);
    }
    
    public ArrayList<Integer> serialize(Node root)
    {
        // code here
        ArrayList<Integer> list = new ArrayList<>();
        inOrder(root, list);
        return list;
    }
    
    public Node createBST(ArrayList<Integer> arr, int s, int e)
    {
        if(s > e)
        return null;
        
        int mid = (s + e) / 2;
        
        Node root = new Node(arr.get(mid));
        root.left = createBST(arr, s, mid-1);
        root.right = createBST(arr, mid+1, e);
        
        return root;
    }

    // Function to deserialize a list and construct the tree.
    public Node deSerialize(ArrayList<Integer> arr)
    {
        // code here
        return createBST(arr, 0, arr.size()-1); 
    }
};
